File: OneToOneCallSample/app/src/main/java/com/stringee/apptoappcallsample/stringee/activity/CallActivity.java
package com.stringee.apptoappcallsample.stringee.activity;

import android.content.Intent;
import android.media.projection.MediaProjectionManager;
import android.os.Build;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.WindowManager;
import android.widget.FrameLayout;

import androidx.activity.OnBackPressedCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.stringee.apptoappcallsample.R;
import com.stringee.apptoappcallsample.databinding.ActivityVideoCallBinding;
import com.stringee.apptoappcallsample.databinding.ActivityVoiceCallBinding;
import com.stringee.apptoappcallsample.databinding.LayoutIncomingCallBinding;
import com.stringee.apptoappcallsample.stringee.common.CallConstants;
import com.stringee.apptoappcallsample.stringee.common.CallEngine;
import com.stringee.apptoappcallsample.stringee.common.CallPermissions;
import com.stringee.apptoappcallsample.stringee.common.CallStatus;
import com.stringee.apptoappcallsample.stringee.listener.CallUiListener;
import com.stringee.apptoappcallsample.stringee.manager.CallSession;
import com.stringee.apptoappcallsample.stringee.manager.StringeeCallManager;
import com.stringee.apptoappcallsample.stringee.service.MediaProjectionCallService;
import com.stringee.common.StringeeAudioManager.AudioDevice;
import com.stringee.video.StringeeVideoTrack;

import org.webrtc.RendererCommon;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/** Permission-aware in-call UI for voice, video, and Call2 screen sharing. */
public class CallActivity extends AppCompatActivity implements View.OnClickListener,
        CallUiListener {
    private ActivityVideoCallBinding videoBinding;
    private ActivityVoiceCallBinding voiceBinding;
    private LayoutIncomingCallBinding incomingBinding;
    private CallSession session;
    private boolean videoCall;
    private boolean permissionFlowCompleted;
    private boolean answerRequested;
    private ActivityResultLauncher<Intent> mediaProjectionLauncher;
    private final List<StringeeVideoTrack> screenTracks = new ArrayList<>();
    private StringeeVideoTrack displayedScreenTrack;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        StringeeCallManager manager = StringeeCallManager.getInstance(this);
        session = manager.getSession();
        long requestedGeneration = getIntent().getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1);
        if (session == null || (requestedGeneration >= 0
                && !manager.ownsSessionGeneration(requestedGeneration))) {
            finish();
            return;
        }
        answerRequested = getIntent().getBooleanExtra(CallConstants.EXTRA_ANSWER, false);
        videoCall = session.isVideoCall();
        videoBinding = ActivityVideoCallBinding.inflate(getLayoutInflater());
        voiceBinding = ActivityVoiceCallBinding.inflate(getLayoutInflater());
        setContentView(videoCall ? videoBinding.getRoot() : voiceBinding.getRoot());
        incomingBinding = videoCall ? videoBinding.vIncomingCall : voiceBinding.vIncomingCall;

        configureWindow();
        bindActions();
        configureInitialUi();
        mediaProjectionLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(), result -> {
                    if (result.getResultCode() == RESULT_OK && result.getData() != null) {
                        Intent serviceIntent = new Intent(this,
                                MediaProjectionCallService.class)
                                .setAction(CallConstants.ACTION_START_IN_CALL)
                                .putExtra(MediaProjectionCallService.EXTRA_PERMISSION_DATA,
                                        result.getData());
                        ContextCompat.startForegroundService(this, serviceIntent);
                    }
                });
        session.registerUiListener(this);
        ensureCallPermissions();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        long requestedGeneration = intent.getLongExtra(
                CallConstants.EXTRA_SESSION_GENERATION, -1);
        if (requestedGeneration >= 0 && !StringeeCallManager.getInstance(this)
                .ownsSessionGeneration(requestedGeneration)) {
            return;
        }
        if (intent.getBooleanExtra(CallConstants.EXTRA_ANSWER, false)) {
            answerRequested = true;
            startCallAfterPermissions();
        }
    }

    private void configureWindow() {
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                | WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            setShowWhenLocked(true);
            setTurnScreenOn(true);
        }
        getOnBackPressedDispatcher().addCallback(this, new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
            }
        });
    }

    private void bindActions() {
        incomingBinding.btnAnswer.setOnClickListener(this);
        incomingBinding.btnReject.setOnClickListener(this);
        if (videoCall) {
            videoBinding.btnEnd.setOnClickListener(this);
            videoBinding.btnMute.setOnClickListener(this);
            videoBinding.btnCamera.setOnClickListener(this);
            videoBinding.btnSwitch.setOnClickListener(this);
            videoBinding.btnShare.setOnClickListener(this);
        } else {
            voiceBinding.btnEnd.setOnClickListener(this);
            voiceBinding.btnMute.setOnClickListener(this);
            voiceBinding.btnSpeaker.setOnClickListener(this);
        }
    }

    private void configureInitialUi() {
        boolean incoming = session.getStatus() == CallStatus.INCOMING;
        incomingBinding.getRoot().setVisibility(incoming ? View.VISIBLE : View.GONE);
        if (videoCall) {
            videoBinding.vInCall.setVisibility(incoming ? View.GONE : View.VISIBLE);
            videoBinding.vShareBtn.setVisibility(
                    session.getEngine() == CallEngine.STRINGEE_CALL2 ? View.VISIBLE : View.GONE);
        } else {
            voiceBinding.vInCall.setVisibility(incoming ? View.GONE : View.VISIBLE);
            voiceBinding.tvUser1.setText(session.getFrom());
        }
        incomingBinding.tvUser.setText(session.getFrom());
    }

    private void ensureCallPermissions() {
        if (CallPermissions.hasCallPermissions(this, videoCall)) {
            permissionFlowCompleted = true;
            startCallAfterPermissions();
        } else {
            CallPermissions.requestCallPermissions(this, videoCall);
        }
    }

    private void startCallAfterPermissions() {
        if (!permissionFlowCompleted || session == null) {
            return;
        }
        if (session.isIncoming()) {
            if (answerRequested) {
                answerRequested = false;
                session.answer();
            }
        } else {
            StringeeCallManager.getInstance(this).startPreparedOutgoing();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode != CallPermissions.REQUEST_CALL_PERMISSIONS) {
            return;
        }
        permissionFlowCompleted = CallPermissions.verify(grantResults);
        if (permissionFlowCompleted) {
            startCallAfterPermissions();
        } else if (session != null) {
            if (session.isIncoming()) {
                session.reject();
            } else {
                session.release();
            }
            finishCallActivity();
        }
    }

    @Override
    public void onClick(View view) {
        if (session == null) {
            return;
        }
        int id = view.getId();
        if (id == R.id.btn_answer) {
            answerRequested = true;
            startCallAfterPermissions();
        } else if (id == R.id.btn_reject) {
            session.reject();
        } else if (id == R.id.btn_end) {
            session.hangUp();
        } else if (id == R.id.btn_mute) {
            session.toggleMute();
        } else if (id == R.id.btn_speaker) {
            session.toggleSpeaker();
        } else if (id == R.id.btn_camera) {
            session.toggleVideo();
        } else if (id == R.id.btn_switch) {
            session.switchCamera();
        } else if (id == R.id.btn_share) {
            if (session.isSharing()) {
                session.stopScreenShare();
            } else {
                session.prepareScreenShare(this, mediaProjectionLauncher,
                        getSystemService(MediaProjectionManager.class));
            }
        }
    }

    @Override
    public void onCallStatus(CallStatus status) {
        runOnUiThread(() -> {
            if (!videoCall) {
                voiceBinding.tvStatus.setText(status.getValue());
            }
            boolean incoming = status == CallStatus.INCOMING;
            incomingBinding.getRoot().setVisibility(incoming ? View.VISIBLE : View.GONE);
            if (videoCall) {
                videoBinding.vInCall.setVisibility(incoming ? View.GONE : View.VISIBLE);
            } else {
                voiceBinding.vInCall.setVisibility(incoming ? View.GONE : View.VISIBLE);
            }
            if (status == CallStatus.ENDED || status == CallStatus.BUSY) {
                finishCallActivity();
            }
        });
    }

    @Override
    public void onError(String message) {
        runOnUiThread(this::finishCallActivity);
    }

    @Override
    public void onLocalVideoAvailable() {
        runOnUiThread(() -> renderView(videoBinding.vLocal, session.getLocalView(), true));
    }

    @Override
    public void onRemoteVideoAvailable() {
        runOnUiThread(() -> {
            if (displayedScreenTrack == null) {
                renderView(videoBinding.vRemote, session.getRemoteView(), false);
            }
        });
    }

    private void renderView(FrameLayout parent, View child, boolean local) {
        if (!videoCall || child == null || session == null) {
            return;
        }
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT);
        params.gravity = Gravity.CENTER;
        parent.removeAllViews();
        parent.addView(child, params);
        if (local) {
            session.renderLocalView();
        } else {
            session.renderRemoteView();
        }
    }

    @Override
    public void onAudioDeviceChanged(AudioDevice selected, Set<AudioDevice> available) {
        if (videoCall) {
            return;
        }
        runOnUiThread(() -> {
            boolean speaker = selected == AudioDevice.SPEAKER_PHONE;
            voiceBinding.btnSpeaker.setBackgroundResource(speaker
                    ? R.drawable.btn_ic_selector : R.drawable.btn_ic_selected_selector);
            voiceBinding.btnSpeaker.setImageResource(speaker
                    ? R.drawable.ic_speaker_on : R.drawable.ic_speaker_off);
        });
    }

    @Override
    public void onMicChanged(boolean enabled) {
        runOnUiThread(() -> {
            if (videoCall) {
                videoBinding.btnMute.setBackgroundResource(enabled
                        ? R.drawable.btn_ic_selected_selector : R.drawable.btn_ic_selector);
                videoBinding.btnMute.setImageResource(enabled
                        ? R.drawable.ic_mic_on : R.drawable.ic_mic_off);
            } else {
                voiceBinding.btnMute.setBackgroundResource(enabled
                        ? R.drawable.btn_ic_selected_selector : R.drawable.btn_ic_selector);
                voiceBinding.btnMute.setImageResource(enabled
                        ? R.drawable.ic_mic_on : R.drawable.ic_mic_off);
            }
        });
    }

    @Override
    public void onVideoChanged(boolean enabled) {
        if (videoCall) {
            runOnUiThread(() -> {
                videoBinding.btnCamera.setBackgroundResource(enabled
                        ? R.drawable.btn_ic_selected_selector : R.drawable.btn_ic_selector);
                videoBinding.btnCamera.setImageResource(enabled
                        ? R.drawable.ic_cam_on : R.drawable.ic_cam_off);
            });
        }
    }

    @Override
    public void onSharingChanged(boolean sharing) {
        if (videoCall) {
            runOnUiThread(() -> {
                videoBinding.btnShare.setBackgroundResource(sharing
                        ? R.drawable.btn_ic_selector : R.drawable.btn_ic_selected_selector);
                videoBinding.btnShare.setImageResource(sharing
                        ? R.drawable.ic_share_off : R.drawable.ic_share);
            });
        }
    }

    @Override
    public void onTimer(String duration) {
        runOnUiThread(() -> {
            if (videoCall) {
                videoBinding.tvTime.setText(duration);
            } else {
                voiceBinding.tvTime.setText(duration);
            }
        });
    }

    @Override
    public void onScreenTrackAdded(StringeeVideoTrack track) {
        runOnUiThread(() -> {
            if (!screenTracks.contains(track)) {
                screenTracks.add(track);
            }
            displayScreenTrack(track);
        });
    }

    @Override
    public void onScreenTrackRemoved(StringeeVideoTrack track) {
        runOnUiThread(() -> {
            screenTracks.removeIf(candidate -> candidate.getId().equals(track.getId()));
            displayedScreenTrack = screenTracks.isEmpty() ? null
                    : screenTracks.get(screenTracks.size() - 1);
            if (displayedScreenTrack == null) {
                renderView(videoBinding.vRemote, session.getRemoteView(), false);
            } else {
                displayScreenTrack(displayedScreenTrack);
            }
        });
    }

    private void displayScreenTrack(StringeeVideoTrack track) {
        if (!videoCall || track == null) {
            return;
        }
        displayedScreenTrack = track;
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT);
        params.gravity = Gravity.CENTER;
        videoBinding.vRemote.removeAllViews();
        videoBinding.vRemote.addView(track.getView2(this), params);
        track.renderView2(RendererCommon.ScalingType.SCALE_ASPECT_FIT);
    }

    private void finishCallActivity() {
        clearVideoViews();
        if (session != null) {
            session.unregisterUiListener(this);
        }
        getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                | WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);
        finish();
    }

    @Override
    protected void onDestroy() {
        clearVideoViews();
        if (session != null) {
            session.unregisterUiListener(this);
        }
        super.onDestroy();
    }

    private void clearVideoViews() {
        if (!videoCall || videoBinding == null) {
            return;
        }
        videoBinding.vLocal.removeAllViews();
        videoBinding.vRemote.removeAllViews();
        screenTracks.clear();
        displayedScreenTrack = null;
    }
}

File: Kotlin-OneToOneCallSample/app/src/main/java/com/stringee/kotlin_onetoonecallsample/stringee/activity/CallActivity.kt
package com.stringee.kotlin_onetoonecallsample.stringee.activity

import android.content.Intent
import android.media.projection.MediaProjectionManager
import android.os.Build
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.view.WindowManager
import android.widget.FrameLayout
import androidx.activity.OnBackPressedCallback
import androidx.activity.addCallback
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.stringee.common.StringeeAudioManager.AudioDevice
import com.stringee.kotlin_onetoonecallsample.R
import com.stringee.kotlin_onetoonecallsample.databinding.ActivityVideoCallBinding
import com.stringee.kotlin_onetoonecallsample.databinding.ActivityVoiceCallBinding
import com.stringee.kotlin_onetoonecallsample.databinding.LayoutIncomingCallBinding
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallConstants
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallEngine
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallPermissions
import com.stringee.kotlin_onetoonecallsample.stringee.common.CallStatus
import com.stringee.kotlin_onetoonecallsample.stringee.listener.CallUiListener
import com.stringee.kotlin_onetoonecallsample.stringee.manager.CallSession
import com.stringee.kotlin_onetoonecallsample.stringee.manager.StringeeCallManager
import com.stringee.kotlin_onetoonecallsample.stringee.service.MediaProjectionCallService
import com.stringee.video.StringeeVideoTrack
import org.webrtc.RendererCommon

/** Permission-aware in-call UI for voice, video, and Call2 screen sharing. */
class CallActivity : AppCompatActivity(), View.OnClickListener, CallUiListener {
    private var videoBinding: ActivityVideoCallBinding? = null
    private var voiceBinding: ActivityVoiceCallBinding? = null
    private var incomingBinding: LayoutIncomingCallBinding? = null
    private var session: CallSession? = null
    private var videoCall = false
    private var permissionFlowCompleted = false
    private var answerRequested = false
    private var mediaProjectionLauncher: ActivityResultLauncher<Intent>? = null
    private val screenTracks: MutableList<StringeeVideoTrack?> = ArrayList<StringeeVideoTrack?>()
    private var displayedScreenTrack: StringeeVideoTrack? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val manager: StringeeCallManager = StringeeCallManager.Companion.getInstance(this)
        session = manager.session
        val requestedGeneration = getIntent().getLongExtra(
            CallConstants.EXTRA_SESSION_GENERATION, -1
        )
        if (session == null || (requestedGeneration >= 0
                    && !manager.ownsSessionGeneration(requestedGeneration))
        ) {
            finish()
            return
        }
        answerRequested = getIntent().getBooleanExtra(CallConstants.EXTRA_ANSWER, false)
        videoCall = session!!.isVideoCall
        videoBinding = ActivityVideoCallBinding.inflate(getLayoutInflater())
        voiceBinding = ActivityVoiceCallBinding.inflate(getLayoutInflater())
        setContentView(if (videoCall) videoBinding!!.getRoot() else voiceBinding!!.getRoot())
        incomingBinding =
            if (videoCall) videoBinding!!.vIncomingCall else voiceBinding!!.vIncomingCall

        configureWindow()
        bindActions()
        configureInitialUi()
        mediaProjectionLauncher = registerForActivityResult(
            StartActivityForResult(), ActivityResultCallback { result: ActivityResult ->
                if (result.resultCode == RESULT_OK && result.data != null) {
                    val serviceIntent = Intent(
                        this,
                        MediaProjectionCallService::class.java
                    )
                        .setAction(CallConstants.ACTION_START_IN_CALL)
                        .putExtra(
                            MediaProjectionCallService.Companion.EXTRA_PERMISSION_DATA,
                            result.data
                        )
                    ContextCompat.startForegroundService(this, serviceIntent)
                }
            })
        session!!.registerUiListener(this)
        ensureCallPermissions()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        val requestedGeneration = intent.getLongExtra(
            CallConstants.EXTRA_SESSION_GENERATION, -1
        )
        if (requestedGeneration >= 0 && !StringeeCallManager.Companion.getInstance(this)
                .ownsSessionGeneration(requestedGeneration)
        ) {
            return
        }
        if (intent.getBooleanExtra(CallConstants.EXTRA_ANSWER, false)) {
            answerRequested = true
            startCallAfterPermissions()
        }
    }

    private fun configureWindow() {
        getWindow().addFlags(
            (WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                    or WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                    or WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)
        )
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            setShowWhenLocked(true)
            setTurnScreenOn(true)
        }
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
            }
        })
    }

    private fun bindActions() {
        incomingBinding!!.btnAnswer.setOnClickListener(this)
        incomingBinding!!.btnReject.setOnClickListener(this)
        if (videoCall) {
            videoBinding!!.btnEnd.setOnClickListener(this)
            videoBinding!!.btnMute.setOnClickListener(this)
            videoBinding!!.btnCamera.setOnClickListener(this)
            videoBinding!!.btnSwitch.setOnClickListener(this)
            videoBinding!!.btnShare.setOnClickListener(this)
        } else {
            voiceBinding!!.btnEnd.setOnClickListener(this)
            voiceBinding!!.btnMute.setOnClickListener(this)
            voiceBinding!!.btnSpeaker.setOnClickListener(this)
        }
    }

    private fun configureInitialUi() {
        val incoming = session!!.status == CallStatus.INCOMING
        incomingBinding!!.getRoot().setVisibility(if (incoming) View.VISIBLE else View.GONE)
        if (videoCall) {
            videoBinding!!.vInCall.setVisibility(if (incoming) View.GONE else View.VISIBLE)
            videoBinding!!.vShareBtn.setVisibility(
                if (session!!.engine == CallEngine.STRINGEE_CALL2) View.VISIBLE else View.GONE
            )
        } else {
            voiceBinding!!.vInCall.setVisibility(if (incoming) View.GONE else View.VISIBLE)
            voiceBinding!!.tvUser1.text = session!!.from
        }
        incomingBinding!!.tvUser.text = session!!.from
    }

    private fun ensureCallPermissions() {
        if (CallPermissions.hasCallPermissions(this, videoCall)) {
            permissionFlowCompleted = true
            startCallAfterPermissions()
        } else {
            CallPermissions.requestCallPermissions(this, videoCall)
        }
    }

    private fun startCallAfterPermissions() {
        if (!permissionFlowCompleted || session == null) {
            return
        }
        if (session!!.isIncoming) {
            if (answerRequested) {
                answerRequested = false
                session!!.answer()
            }
        } else {
            StringeeCallManager.Companion.getInstance(this).startPreparedOutgoing()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode != CallPermissions.REQUEST_CALL_PERMISSIONS) {
            return
        }
        permissionFlowCompleted = CallPermissions.verify(grantResults)
        if (permissionFlowCompleted) {
            startCallAfterPermissions()
        } else if (session != null) {
            if (session!!.isIncoming) {
                session!!.reject()
            } else {
                session!!.release()
            }
            finishCallActivity()
        }
    }

    override fun onClick(view: View) {
        if (session == null) {
            return
        }
        val id = view.getId()
        if (id == R.id.btn_answer) {
            answerRequested = true
            startCallAfterPermissions()
        } else if (id == R.id.btn_reject) {
            session!!.reject()
        } else if (id == R.id.btn_end) {
            session!!.hangUp()
        } else if (id == R.id.btn_mute) {
            session!!.toggleMute()
        } else if (id == R.id.btn_speaker) {
            session!!.toggleSpeaker()
        } else if (id == R.id.btn_camera) {
            session!!.toggleVideo()
        } else if (id == R.id.btn_switch) {
            session!!.switchCamera()
        } else if (id == R.id.btn_share) {
            if (session!!.isSharing) {
                session!!.stopScreenShare()
            } else {
                session!!.prepareScreenShare(
                    this, requireNotNull(mediaProjectionLauncher),
                    getSystemService(MediaProjectionManager::class.java)
                )
            }
        }
    }

    override fun onCallStatus(status: CallStatus) {
        runOnUiThread(Runnable {
            if (!videoCall) {
                voiceBinding!!.tvStatus.setText(status.value)
            }
            val incoming = status == CallStatus.INCOMING
            incomingBinding!!.getRoot().setVisibility(if (incoming) View.VISIBLE else View.GONE)
            if (videoCall) {
                videoBinding!!.vInCall.setVisibility(if (incoming) View.GONE else View.VISIBLE)
            } else {
                voiceBinding!!.vInCall.setVisibility(if (incoming) View.GONE else View.VISIBLE)
            }
            if (status == CallStatus.ENDED || status == CallStatus.BUSY) {
                finishCallActivity()
            }
        })
    }

    override fun onError(message: String?) {
        runOnUiThread(Runnable { this.finishCallActivity() })
    }

    override fun onLocalVideoAvailable() {
        runOnUiThread(Runnable {
            renderView(
                videoBinding!!.vLocal,
                session!!.localView,
                true
            )
        })
    }

    override fun onRemoteVideoAvailable() {
        runOnUiThread(Runnable {
            if (displayedScreenTrack == null) {
                renderView(videoBinding!!.vRemote, session!!.remoteView, false)
            }
        })
    }

    private fun renderView(parent: FrameLayout, child: View?, local: Boolean) {
        if (!videoCall || child == null || session == null) {
            return
        }
        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT
        )
        params.gravity = Gravity.CENTER
        parent.removeAllViews()
        parent.addView(child, params)
        if (local) {
            session!!.renderLocalView()
        } else {
            session!!.renderRemoteView()
        }
    }

    override fun onAudioDeviceChanged(
        selected: AudioDevice?,
        available: MutableSet<AudioDevice?>?
    ) {
        if (videoCall) {
            return
        }
        runOnUiThread(Runnable {
            val speaker = selected == AudioDevice.SPEAKER_PHONE
            voiceBinding!!.btnSpeaker.setBackgroundResource(
                if (speaker)
                    R.drawable.btn_ic_selector
                else
                    R.drawable.btn_ic_selected_selector
            )
            voiceBinding!!.btnSpeaker.setImageResource(
                if (speaker)
                    R.drawable.ic_speaker_on
                else
                    R.drawable.ic_speaker_off
            )
        })
    }

    override fun onMicChanged(enabled: Boolean) {
        runOnUiThread(Runnable {
            if (videoCall) {
                videoBinding!!.btnMute.setBackgroundResource(
                    if (enabled)
                        R.drawable.btn_ic_selected_selector
                    else
                        R.drawable.btn_ic_selector
                )
                videoBinding!!.btnMute.setImageResource(
                    if (enabled)
                        R.drawable.ic_mic_on
                    else
                        R.drawable.ic_mic_off
                )
            } else {
                voiceBinding!!.btnMute.setBackgroundResource(
                    if (enabled)
                        R.drawable.btn_ic_selected_selector
                    else
                        R.drawable.btn_ic_selector
                )
                voiceBinding!!.btnMute.setImageResource(
                    if (enabled)
                        R.drawable.ic_mic_on
                    else
                        R.drawable.ic_mic_off
                )
            }
        })
    }

    override fun onVideoChanged(enabled: Boolean) {
        if (videoCall) {
            runOnUiThread(Runnable {
                videoBinding!!.btnCamera.setBackgroundResource(
                    if (enabled)
                        R.drawable.btn_ic_selected_selector
                    else
                        R.drawable.btn_ic_selector
                )
                videoBinding!!.btnCamera.setImageResource(
                    if (enabled)
                        R.drawable.ic_cam_on
                    else
                        R.drawable.ic_cam_off
                )
            })
        }
    }

    override fun onSharingChanged(sharing: Boolean) {
        if (videoCall) {
            runOnUiThread(Runnable {
                videoBinding!!.btnShare.setBackgroundResource(
                    if (sharing)
                        R.drawable.btn_ic_selector
                    else
                        R.drawable.btn_ic_selected_selector
                )
                videoBinding!!.btnShare.setImageResource(
                    if (sharing)
                        R.drawable.ic_share_off
                    else
                        R.drawable.ic_share
                )
            })
        }
    }

    override fun onTimer(duration: String?) {
        runOnUiThread(Runnable {
            if (videoCall) {
                videoBinding!!.tvTime.setText(duration)
            } else {
                voiceBinding!!.tvTime.setText(duration)
            }
        })
    }

    override fun onScreenTrackAdded(track: StringeeVideoTrack?) {
        runOnUiThread(Runnable {
            if (!screenTracks.contains(track)) {
                screenTracks.add(track)
            }
            displayScreenTrack(track)
        })
    }

    override fun onScreenTrackRemoved(track: StringeeVideoTrack) {
        runOnUiThread(Runnable {
            screenTracks.removeIf { candidate: StringeeVideoTrack? -> candidate!!.getId() == track.getId() }
            displayedScreenTrack = if (screenTracks.isEmpty())
                null
            else
                screenTracks.get(screenTracks.size - 1)
            if (displayedScreenTrack == null) {
                renderView(videoBinding!!.vRemote, session!!.remoteView, false)
            } else {
                displayScreenTrack(displayedScreenTrack)
            }
        })
    }

    private fun displayScreenTrack(track: StringeeVideoTrack?) {
        if (!videoCall || track == null) {
            return
        }
        displayedScreenTrack = track
        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT
        )
        params.gravity = Gravity.CENTER
        videoBinding!!.vRemote.removeAllViews()
        videoBinding!!.vRemote.addView(track.getView2(this), params)
        track.renderView2(RendererCommon.ScalingType.SCALE_ASPECT_FIT)
    }

    private fun finishCallActivity() {
        clearVideoViews()
        if (session != null) {
            session!!.unregisterUiListener(this)
        }
        getWindow().clearFlags(
            (WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                    or WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                    or WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)
        )
        finish()
    }

    override fun onDestroy() {
        clearVideoViews()
        if (session != null) {
            session!!.unregisterUiListener(this)
        }
        super.onDestroy()
    }

    private fun clearVideoViews() {
        if (!videoCall || videoBinding == null) {
            return
        }
        videoBinding!!.vLocal.removeAllViews()
        videoBinding!!.vRemote.removeAllViews()
        screenTracks.clear()
        displayedScreenTrack = null
    }
}
